doc: add decision docs for decision meeting 15.07.2021 (#3919)
ElektraInitiative/libelektra: doc/decisions/reference_counting.md

# Reference Counting

## Problem

- locking is not reset when ref counting again gets 0 (adding to keyset and
  pop again) #2202
- C++ API for KeySet and Key has unexpected differences: also use ref counting
  for KeySets (also suggested in #1332)

## Constraints

## Assumptions

## Considered Alternatives

- make ref counting thread safe (probably useful for JNI)
- start with 1 for reference counting and let keyDecRef do keyDel

## Decision

> - add second reference counter to Key

Reviewer: See #2202: probably it is simply a documentation issue and no second ref counter is needed

Reviewer 2: Honestly, what was the decision meeting for, if we immediately reverse one of the main decisions?

See https://github.com/ElektraInitiative/libelektra/pull/3949#issuecomment-882484784 for an idea on how documentation/API naming can make the second ref counter much easier to understand.

Reviewer: > Honestly, what was the decision meeting for, if we immediately reverse one of the main decisions?

In some cases we also went into timeouts and agreed there would be a follow up.

> See #3949 (comment) for an idea on how documentation/API naming can make the second ref counter much easier to understand.

I like the idea but this immediately raises the question why locking the value or meta-data does not get ref-counted.

Reviewer 2: > raises the question why locking the value or meta-data does not get ref-counted.

To me the question is: Why is locking value/meta-data a thing at all? Is there a use-case for it?

AFAICT we only directly use `KEY_FLAG_RO_VALUE` and `KEY_FLAG_RO_META` inside `keySetMeta`. `KEY_LOCK_VALUE` and `KEY_LOCK_META` are not used at all.

For the metadata case we should actually check the key name in `keySetRawValue` and `keyMeta`. Otherwise it won't work with `ksAppendKey (keyMeta (k), ...)` or if you create a meta key via `keySetName (k, "meta:/foo")`.

Reviewer: > To me the question is: Why is locking value/meta-data a thing at all? Is there a use-case for it?

It is an invariant of a KeySet that its keys are ordered. The locking ensures this invariant.

> KEY_LOCK_META

The idea was to use this for meta keys. But probably this is now useless, as meta keys can be recognized by the namespace.

Reviewer 2: > It is an invariant of a KeySet that its keys are ordered. The locking ensures this invariant.

ordered **by name**

Locking **value** or **metadata** has nothing to do with this.

Reviewer: Sorry, misread. I was thinking about if we can get rid of locking at all. And probably we can, we could simply disallow name changes with ref counter != initial value.

To answer your previous question: Locking of values was added for completeness. Other than read-only keys there is probably no usecase.

Reviewer 2: > we could simply disallow name changes with ref counter != initial value.

Then you can't ever change the name from Java or another language with automatic garbage collection, where incrementing the ref counter is required.

Reviewer 3: As far as i understand:
* keysets store keys in sorted by key names
* a key's order within a key set is determined on insertion
* the purpose of locking a key's name is to prevent (potentially) compromising the order of keys for keysets the key is contained in.

To better understand the issue:
* what are the use cases for changing a key's name? why not making it immutable?
* why is order within keysets so important? would it be sufficient to ensure order explicitly for use-cases requiring it?

Reviewer 4: > what are the use cases for changing a key's name? why not making it immutable?

In short, to avoid doing some memory allocations. In my case I had a loop where a key was removed and a new one was added with a new basename. The same can be done with immutable keys, but it requires allocating new memory for N keys instead of reusing the old keys.

I was just confused that unlocking is currently not possible at all. When a key is added to a keyset the name is locked, but when it is removed from the keyset the name is not unlocked. To be consistent we should either have immutable keys or provide a way to unlock the keys again.

> why is order within keysets so important? 

It is required because we use binary search to find a key in the keyset. It is also a nice property and I don't see much reason to rework this, unless there are some huge benefits?

Reviewer 2: I think the Hashmap requires the order as well, since it only stores indices into the array. Also `ksCut` relies on the order of the `KeySet` and in general many things are easier or more efficient to implement, if there is a fixed order.

> I was just confused that unlocking is currently not possible at all.

I had the same experience.

> - add second reference counter to Key (size 1 byte)

Reviewer 2: ```suggestion
- add second reference counter to Key
```

I think the size is still TBD. Also it shouldn't really be something that is fixed in the decision.

> - one is for KeySet, the other is for the rest (for external keyset references)

Reviewer 2: ```suggestion
- One counter is for locking references, the other one for general references. A locking reference automatically locks/unlocks the keyname.
```

- introduce reference counter for KeySets (for external keyset references)

> - keyIncRef Fehlercode returnen on MAX_SIZE

Reviewer 2: ```suggestion
- `keyIncRef` (or equivalent) return error in case of reference counter overflow
```

> - `keyIncRef` (or equivalent) return error in case of reference counter overflow

Reviewer: Let us settle for keyBorrow (see #3949) unless someone comes up with a better name (maybe keyShare). The rename should not be too costly, keyIncRef is not heavily used.

- use fixed sized types for reference counters
- increment/decrement references before/after passing instances to plugins

## Rationale

- Adding a second reference counter to Key and reducing the size of both significantly

>   actually saves memory compared to the previous solution.

Reviewer: not really, in #3949 is now 16bit proposed, so at best we stay at the same size.

Reviewer 2: The current ref counter is `size_t`, so on most modern (i.e. 64 bit) machines we safe 4 Bytes, assuming both counters will be 16bit. #3949 actually only proposes a size of `KeySet` not for `Key`.

> - The added complexity of maintaining two reference counters is worth the

Reviewer: This is not an argument, what is the gained functionality?

Reviewer 2: I understood it as: Adding the second ref counter can be done in a way that introduces very little complexity. At the same time (automatically) unlocking keys is a big benefit, because you don't need to make unnecessary copies of keys.

# Reference Counting

## Problem

- locking is not reset when ref counting again gets 0 (adding to keyset and pop again) #2202
- C++ API for KeySet and Key has unexpected differences: also use ref counting for KeySets (also suggested in #1332)

## Constraints

## Assumptions

## Considered Alternatives

- make ref counting thread safe (probably useful for JNI)
- start with 1 for reference counting and let keyDecRef do keyDel (remove ksDel&keyDel; or remove keyDecRef, rename keyIncRef to keyBorrow (or similar) and keyDecRef to keyDel) #1332

## Decision

Instead of renaming `keyIncRef` to `keyBorrow`, I would do this:

```c
/**
  * @return a new reference to k
  */
Key * keyBorrow (Key * k) {
   keyIncRef (k);
   return k;
}
```

`keyIncRef` and `keyDecRef` would be internal/private/static (whatever fits). The same `ks*` functions should exist for `KeySet`.

## Rationale

The end effect is basically the same. Users should

- use `keyNew` and `keyCopy`/`keyDup` to create entirely new `Key *` that don't have any links to other `Key *`.
- use `keyBorrow` to create a new `Key *` that is a reference an existing `Key *`.

> - use `keyDel` to dispose of a `Key *` (whether the `Key *` comes from `keyNew`, `keyCopy`/`keyDup` or `keyBorrow`).

Reviewer: At the moment keyDel does not decrement the counter, so how to use keyDel to get rid of a key from keyBorrow?

## Implications

> 

Reviewer: Implications should state which code needs to be changed.

Reviewer: Still missing.

Reviewer: Still missing.
